homeassistant/components/ecovacs/__init__.py

```diff
@@ -69,28 +71,32 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
             device.get("did"),
             device.get("nick"),
         )
-        vacbot = VacBot(
-            ecovacs_api.uid,
-            ecovacs_api.REALM,
-            ecovacs_api.resource,
-            ecovacs_api.user_access_token,
-            device,
-            config[DOMAIN].get(CONF_CONTINENT).lower(),
-            monitor=True,
+        vacbot = await hass.async_add_executor_job(
+            partial(
+                VacBot,
+                ecovacs_api.uid,
+                ecovacs_api.REALM,
+                ecovacs_api.resource,
+                ecovacs_api.user_access_token,
+                device,
+                config[DOMAIN].get(CONF_CONTINENT).lower(),
+                monitor=True,
+            )
         )
         hass.data[ECOVACS_DEVICES].append(vacbot)
 
-    def stop(event: object) -> None:
+    async def async_stop(event: object) -> None:
         """Shut down open connections to Ecovacs XMPP server."""
-        for device in hass.data[ECOVACS_DEVICES]:
+        devices: list[VacBot] = hass.data[ECOVACS_DEVICES]
+        for device in devices:
             _LOGGER.info(
                 "Shutting down connection to Ecovacs device %s",
                 device.vacuum.get("did"),
             )
-            device.disconnect()
+            await hass.async_add_executor_job(device.disconnect)
 
     # Listen for HA stop to disconnect.
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_stop)
 
     if hass.data[ECOVACS_DEVICES]:
         _LOGGER.debug("Starting vacuum components")
```

Comment:
We need to use `discovery.async_load_platform` below. We may need to create a task for those calls. Nor sure if there's still a risk for deadlock nowadays.